ipa-core/src/protocol/hybrid/agg.rs

```diff
@@ -0,0 +1,235 @@
+    for report in reports {
+        let match_key = report.match_key;
+        let entry = reports_by_matchkey.entry(match_key).or_insert((
+            [
+                AggregateableHybridReport::<BK, V>::ZERO,
+                AggregateableHybridReport::<BK, V>::ZERO,
+            ],
+            0,
+        ));
+        if entry.1 == 0 {
+            // If the count is 0, replace the first element with the new report
+            entry.0[0] = report.into();
+            entry.1 += 1;
+        } else if entry.1 == 1 {
+            // If the count is 1, replace the second element with the new report
+            entry.0[1] = report.into();
+            entry.1 += 1;
+        } else {
+            // If the count is 2 or more, increment the counter and drop the report
+            entry.1 += 1;
+        }
+    }
+
+    // we only keep the reports from match_keys that provided exactly 2 reports
+    let report_pairs: Vec<[AggregateableHybridReport<BK, V>; 2]> = reports_by_matchkey
+        .into_iter()
+        .filter_map(|(_, v)| if v.1 == 2 { Some(v.0) } else { None })
+        .collect::<Vec<_>>();
+
+    // let mut agg_reports = Vec::with_capacity(report_pairs.len());
+
+    let chunk_size = TARGET_PROOF_SIZE;
+
+    let dzkp_validator = ctx.clone().dzkp_validator(
+        MaliciousProtocolSteps {
+            protocol: &HybridStep::GroupBySum,
+            validate: &HybridStep::GroupBySumValidate,
+        },
+        std::cmp::min(ctx.active_work().get(), chunk_size.next_power_of_two()),
+    );
+
+    let ctx = dzkp_validator
+        .context()
+        .set_total_records(TotalRecords::specified(2 * report_pairs.len())?);
+
+    let agg_work = stream::iter(report_pairs)
+        .enumerate()
+        .map(|(idx, reports)| {
+            let record_id_bk = RecordId::FIRST + 2 * idx;
+            let record_id_v = RecordId::FIRST + 2 * idx + 1;
+            let agg_ctx = ctx.clone();
+            async move {
+                let (breakdown_key, _) = integer_add::<_, EightBitStep, 1>(
+                    agg_ctx.narrow(&AggregateReportsStep::Add),
+                    record_id_bk,
+                    &reports[0].breakdown_key.to_bits(),
+                    &reports[1].breakdown_key.to_bits(),
+                )
+                .await?;
+                let (value, _) = integer_add::<_, ThirtyTwoBitStep, 1>(
+                    agg_ctx.narrow(&AggregateReportsStep::Add),
+                    record_id_v,
+                    &reports[0].value.to_bits(),
+                    &reports[1].value.to_bits(),
+                )
+                .await?;
+                Ok::<_, Error>(AggregateableHybridReport::<BK, V> {
+                    match_key: (),
+                    breakdown_key: breakdown_key.collect_bits(),
+                    value: value.collect_bits(),
+                })
+            }
+        });
+
+    let agg_result = seq_join(ctx.active_work(), agg_work)
+        .try_collect::<Vec<_>>()
+        .await?;
```

Comment:
@akoshelev is this the right way to use `seq_join`? it compiles, but it's still stalling.

Reply:
I don't think you need to do record id math here, just use `idx` as record id. Where does it stall?

Reply:
> I don't think you need to do record id math here, just use idx as record id.

is it ok to use the same record_id for both additions? I figured they needed to be distinct.

Reply:
here are the logs where it stalls:
```
2024-11-13T23:32:26.777285Z TRACE {role=H2}:receive{i=0 from=H3 gate="protocol/iter000/group_by_sum/add/bit0"}:send_stream{to=H2 gate=gate=protocol/iter000/group_by_sum/add/bit0}: ipa_core::helpers::gateway::send: new
2024-11-13T23:32:26.777336Z TRACE {role=H2}:receive{i=0 from=H3 gate="protocol/iter000/group_by_sum/add/bit0"}:send_stream{to=H2 gate=gate=protocol/iter000/group_by_sum/add/bit0}: ipa_core::helpers::gateway::send: close time.busy=3.25µs time.idle=47.2µs
2024-11-13T23:32:26.777381Z TRACE {role=H3}:receive{i=0 from=H1 gate="protocol/iter000/group_by_sum/add/bit0"}:send_stream{to=H3 gate=gate=protocol/iter000/group_by_sum/add/bit0}: ipa_core::helpers::gateway::send: new
2024-11-13T23:32:26.777475Z TRACE {role=H3}:receive{i=0 from=H1 gate="protocol/iter000/group_by_sum/add/bit0"}:send_stream{to=H3 gate=gate=protocol/iter000/group_by_sum/add/bit0}: ipa_core::helpers::gateway::send: close time.busy=2.75µs time.idle=90.9µs
2024-11-13T23:32:36.771245Z  WARN stall_detector{role=H2}: ipa_core::helpers::gateway::stall_detection::gateway: Helper is stalled sn=4 state=
{
"gate=protocol/iter000/group_by_sum/add/bit0", from=H3. Waiting to receive records ["[0, 1]"].
}
2024-11-13T23:32:36.771278Z  WARN stall_detector{role=H1}: ipa_core::helpers::gateway::stall_detection::gateway: Helper is stalled sn=4 state=
{
"gate=protocol/iter000/group_by_sum/add/bit0", from=H2. Waiting to receive records ["[0, 1]"].
}
2024-11-13T23:32:36.771271Z  WARN stall_detector{role=H3}: ipa_core::helpers::gateway::stall_detection::gateway: Helper is stalled sn=4 state=
{
"gate=protocol/iter000/group_by_sum/add/bit0", from=H1. Waiting to receive records ["[0, 1]"].
```